File: 0994-rotting-oranges/0994-rotting-oranges.java
class Pair {
    public int r;
    public int c;

    public Pair(int r, int c) {
        this.r = r;
        this.c = c;
    }
}

class Solution {
    public int orangesRotting(int[][] grid) {
        int[] dr = {0, 0, -1, 1};
        int[] dc = {1, -1, 0, 0};

        boolean[][] chk = new boolean[grid.length][grid[0].length];
        Queue<Pair> que = new ArrayDeque<>();
        int result = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 2) {
                    que.add(new Pair(i, j));
                    chk[i][j] = true;
                }
            }
        }

        while (!que.isEmpty()) {
            int n = que.size();
            boolean flag = false;

            for (int i = 0; i < n; i++) {
                int r = que.peek().r, c = que.peek().c;
                que.poll();
                for (int j = 0; j < 4; j++) {
                    int nr = r + dr[j], nc = c + dc[j];
                    if (nr < 0 || nc < 0 ||
                        nr >= grid.length || nc >= grid[0].length ||
                        grid[nr][nc] != 1 || chk[nr][nc]) continue;
                    grid[nr][nc] = 2;
                    que.add(new Pair(nr, nc));
                    chk[nr][nc] = true;
                    flag = true;
                }
            }

            if (flag) result++;
        }

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) return -1;
            }
        }

        return result;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        dr = [0, 0, -1, 1]
        dc = [1, -1, 0, 0]

        chk = [[False for i in range(len(grid[0]))] for j in range(len(grid))]
        que = deque()
        result = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    que.append((i, j))
                    chk[i][j] = True

        while que:
            n = len(que)
            flag = False
            for i in range(n):
                r, c = que.popleft()
                for i in range(4):
                    nr, nc = r + dr[i], c + dc[i]
                    if (nr < 0 or nc < 0 or
                        nr >= len(grid) or nc >= len(grid[0]) or
                        grid[nr][nc] != 1 or
                        chk[nr][nc]):
                        continue
                    grid[nr][nc] = 2
                    que.append((nr, nc))
                    chk[nr][nc] = True
                    flag = True
            if flag: result += 1

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    return -1

        return result
